feat(Item): create requirements and Item class

brainbackdoor opened the pull request.

Commit:
docs(README.md): create requirements and term dictionaries.

Commit:
feature(item): create item class

etff committed:
Merge pull request #1 from brainbackdoor/feature/glossary

docs(README.md): create requirements and term dictionaries.

brainbackdoor committed:
feature(Seller): create Seller class

etff committed:
Merge pull request #2 from brainbackdoor/feature/seller

feature(Seller): create Seller class

brainbackdoor committed:
feature(ItemService): implement search function

etff committed:
Merge pull request #3 from brainbackdoor/feature/search

feature(ItemService): implement search function

brainbackdoor committed:
feature(Dollar): create Dollar class

etff committed:
Merge pull request #4 from brainbackdoor/feature/dollar

feature(Dollar): create Dollar class

daeheejeong committed:
docs(README.md): update requirements

ivvve committed:
docs(README.md): add currency at 'UnitPrice' in glossary

ivvve committed:
feature(Item): refactor Item class

etff committed:
Merge pull request #5 from brainbackdoor/feature/item

feature(Item): refactor Item class

etff committed:
Merge pull request #6 from brainbackdoor/feature/glossary

Feature/glossary

wotjd243 commented on README.md:
```diff
@@ -1 +1,56 @@
+- 쇼핑몰의 상품 조회 주문을 ddd로 구현한다.
+
+- 판매자
+    - [ ]  eBay에 물품을 등록하여 판매하는 사람
```
"eBay에 물품을 등록하여 판매하는 사람"은 eBay에 실제 물품을 등록하는 것인가요? 아니면 eBay에 등록된 물품을 파는 것인가요?

wotjd243 commented on src/main/java/io/github/wotjd243/support/ecommerce/product/ItemRepository.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public static List<Item> findByKeywords(String keywords) {
+        return items.stream().filter(item -> item.match(keywords)).collect(Collectors.toList());
```
"한 개 이상의 단어로 이루어진 검색 단어. 검색어를 이용해 상품정보를 가져온다. 최소 3글자 이상을 사용한다."에 대한 값 검증이 있으면 좋을 것 같아요.

wotjd243 commented on src/main/java/io/github/wotjd243/support/ecommerce/product/ItemRepository.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ItemRepository {
```
REPOSITORY는 AGGREGATE 단위로 들고 오도록 구현하는 것이 불완전한 AGGREGATE를 들고 오는 것을 방지할 수 있습니다.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/product/domain/Item.java:
```diff
@@ -0,0 +1,65 @@
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class Item {
```
설계의 "Item은 고유한 ID를 갖는다."가 구현되지 않았습니다. 설계 또는 구현을 바꿀 필요가 있어 보여요.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/product/domain/Dollar.java:
```diff
@@ -0,0 +1,17 @@
+package io.github.wotjd243.ecommerce.product.domain;
+
+public class Dollar {
```
VALUE OBJECT 👍

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/product/domain/Product.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public List<Item> getItems() {
+        return items;
```
`items`에 대한 주솟값을 그대로 넘겨주게 되면 외부에 의한 변조가 일어나 `Product` AGGREGATE를 해칠 수 있습니다. 새로운 리스트 컬렉션에 담아 외부에 의한 변조를 막는 것은 어떨까요?

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/product/domain/Item.java:
```diff
@@ -0,0 +1,65 @@
+                '}';
+    }
+
+    enum SellingState {
```
Enum 👍

wotjd243 commented on src/test/java/io/github/wotjd243/ecommerce/product/application/ItemServiceTest.java:
```diff
@@ -0,0 +1,20 @@
+        String keywords = "DDD";
+        List<Item> items = service.search(keywords);
+
+        assertThat(items.get(0).isSamePrice(25.0)).isTrue();
```
테스트 명과 테스트 코드가 매칭될 수 있게 연관 지어 작성하는 것은 어떨까요?

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/user/domain/Seller.java:
```diff
@@ -0,0 +1,22 @@
+        this.id = id;
+    }
+
+    public void registerGoods(Product product, Item item) {
```
`Seller` AGGREGATE에 `Product` AGGREGATE를 인자로 넘길 경우, `Product` AGGREGATE가 오용될 소지가 있습니다. SERVICE를 이용해보는 것이 어떨까요?

wotjd243 reviewed (changes requested):
테스트 코드 작성과 적절한 VALUE OBJECT 사용이 좋았습니다. 👍 
피드백 남겼으니 확인하고 개선할 부분 있으면 반영해 보세요.
코드 개선 시, 모델링도 함께 업데이트되어야 하는 점 잊지 마세요.

ivvve committed:
feature(Search): add SearchService in 'item' (#7)

* feature(Search): add SearchService in 'item'

etff committed:
feature(OrderService): implement order function (#8)

우선 merge하고 수정 진행하겠습니다.

daeheejeong committed:
feat(shippingaddress) (#10)

* create entity ShippingAddress, ShippingAddress dto and UserService

* refactor (shipping address): ZipCode Entity, ShippingAddressService, ShippingAddresses

brainbackdoor committed:
refactor for step1 feedback (#9)

* refactor: README.md 파일의 요구사항 수정 및 package 구조 변경

* feat(ItemService): add feature of register item

* feat(Order): add feature of sum price

* feat(OrderService): add feature of order

* feat(User): add feature of register user

* feat(User): find user's registered items.

* refactor: Delete lombok dependencies

ivvve committed:
Feature/item (#11)

* feature(Stock): create Stock

* feature(Stock): create ItemState

* feature(Page): implement paging, sorting

* feature(Item): implement item repository, domain feature, service

etff committed:
add shopping basket (#12)

daeheejeong committed:
refactor (shipping address) (#13)

Apply reviewed contents
- remove ShippingAddress service & repository
- add ShippingAddress first class collection (list -> set)
- add ZipCode validation

wotjd243 commented on bin/main/io/github/wotjd243/ecommerce/item/application/dto/ItemDto.kt:
```diff
@@ -0,0 +1,14 @@
+package io.github.wotjd243.ecommerce.item.application.dto
+
+data class ItemRequestDto(
```
코틀린 데이터 클래스 👍

wotjd243 commented on README.md:
```diff
@@ -1 +1,77 @@
+- 고려중인물품
+    - [X] 고려중인 물품은 물품의 이름, 물품의 단가, 이미지주소와 수량으로 구성되어 있다.
+
+## 용어 사전
```
`Stock`이라는 것은 물품의 수량을 뜻할까요?

ivvve replied:
Stock은 `재고`를 의미합니다.
용어사전에 업데이트가 필요하겠네요.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/item/domain/Item.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public ItemResponseDto toDto() {
+        return new ItemResponseDto(detail.getTitle(), detail.getPrice(), detail.getGalleryUrl(), getStock(), itemState.name());
```
도메인 계층인 `Item`이 응용 계층인 `ItemResponseDto`를 의존하고 있네요. 만약 응용 계층의 요구 사항(표현 영역에서 받고자 하는 값의 형태 등)이 바뀐다면 도메인 계층도 그것에 영향을 받지 않을까요? 도메인 주도 설계의 핵심은 외부의 영향을 받지 않는 도메인 계층을 만드는 데에 있습니다.

ivvve replied:
좋은 지적 감사합니다.👍
해당 내용 적용하도록 하겠습니다

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/item/domain/ItemRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+    List<Item> findAll();
+
+    List<Item> findAll(PagingDto paging);
```
마찬가지로 응용 계층의 영향을 받고 있네요.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/item/domain/Stock.java:
```diff
@@ -0,0 +1,36 @@
+        validateStock(numberOfDecrement);
+
+        this.value -= numberOfDecrement;
+        return this;
```
VALUE OBJECT는 불변이 보장되어야 하며 밸류 객체의 값을 변경하는 방법은 새로운 밸류 객체를 할당하는 것뿐입니다. 그렇다면 재고의 수량을 줄인 새로운 `Stock` VALUE OBJECT가 반환되어야 하지 않을까요?

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/order/application/OrderService.java:
```diff
@@ -0,0 +1,52 @@
+public class OrderService {
+    private final OrderRepository orderRepository;
+    private final UserService userService;
+    private boolean isPaid;
```
Spring Bean의 기본 Scope는 `singleton`이며, Singleton이 멀티 스레드 환경에서 서비스 형태로 사용되는 경우에는 상태 정보를 가지고 있지 않은 무상태(stateless) 방식으로 만들어져야 thread-safe 할 수 있습니다. 여러 클라이언트가 동시에 `isPaid`에 접근하여 값을 변경한다면 어떤 일이 발생할까요?

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/order/domain/Order.java:
```diff
@@ -0,0 +1,44 @@
+        this.basket = basket;
+    }
+
+    public OrderResponseDto toDto() {
```
도메인 계층에서 응용 계층으로 의존하는 코드입니다.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/order/domain/ShoppingBasket.java:
```diff
@@ -0,0 +1,40 @@
+    public ShoppingBasket(Buyer buyer, List<ConsideringItem> items) {
+        this.id = UUID.randomUUID().toString();
+        this.buyer = buyer;
+        this.items = items;
```
컬렉션의 주솟값을 그대로 넘겨주게 되면 외부에 의한 변조가 일어나 불변성을 해칠 수 있어요. 새로운 컬렉션에 담아 외부에 의한 변조를 막는 것이 좋아 보여요.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/user/domain/ShippingAddresses.java:
```diff
@@ -0,0 +1,19 @@
+
+public class ShippingAddresses {
+
+    private final Set<ShippingAddress> shippingAddresses;
```
일급 컬렉션 사용 👍

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/user/domain/User.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public ShippingAddress addShippingAddress(String address, String zipCode) {
+        return shippingAddresses.add(new ShippingAddress(address, zipCode));
```
`shippingAddresses`의 선언만 되었으며 할당은 되지 않아 `NullPointErexception`이 발생할 수 있습니다.

wotjd243 commented on src/main/java/io/github/wotjd243/ecommerce/user/domain/ZipCode.java:
```diff
@@ -0,0 +1,66 @@
+
+        POSTALCODE("\\d{5}");
+
+        private final String regex;
```
`String`을 패턴 매칭을 위해 `Pattern`으로 바꾸는 것은 꽤나 비싼 행위입니다. 미리 `Pattern`으로 만들어 사용하는 것은 어떨까요?

wotjd243 commented on src/test/java/io/github/wotjd243/ecommerce/user/domain/ShippingAddressesTest.java:
```diff
@@ -0,0 +1,20 @@
+        shippingAddresses.add(new ShippingAddress("서울시 마포구 공덕동", "04208"));
+        assertThatIllegalArgumentException().isThrownBy(
+                () -> shippingAddresses.add(new ShippingAddress("서울시 마포구 공덕동", "04208")
+                ));
```
멋진 VALUE OBJECT와 일급 컬렉션 구현, 테스트 코드 작성입니다. 💯

wotjd243 commented on src/test/java/io/github/wotjd243/ecommerce/order/domain/ShoppingBasketTest.java:
```diff
@@ -0,0 +1,26 @@
+    public void 장바구니에_담은_상품의_총_금액_계산() {
+        ShoppingBasket basket = new ShoppingBasket(BasketUtils.consider(items));
+
+        assertThat(basket.sumPrice()).isEqualTo(251.46);
```
다른 개발자가 이 테스트 코드를 본다면 '251.46'라는 숫자가 어떻게 나온 건지 알 수 있을까요? 테스트를 위해 주어진 상황을 조금 더 친절하게 설명하면 좋을 것 같아요.

wotjd243 commented on README.md:
```diff
@@ -1 +1,77 @@
+- `Item`은 고유한 ID를 갖는다.
+- `Item`은 물품의 이름, 가격, 수량, 상태를 갖는다.
+- `Item`은 다음의 상태값을 갖는다. (BEFORE_SELLING: 판매 전, SELLING: 판매 중, SOLD_OUT: 재고 없음)
+- `Item`은 조회 시 가격순, 이름순으로 정렬을 할 수 있다.
```
현재까지 구현된 것을 모델링으로 표현해보는 것은 어떨까요?

wotjd243 reviewed (approved):
팀원들 간의 코드 리뷰는 저도 항상 잘 보고 있습니다. 👍 
코틀린 데이터 클래스 사용이 굉장히 인상적입니다.
하지만 바이너리 파일이 Git으로 관리되는 점은 수정이 필요해 보여요.
Git으로 프로젝트를 관리할 때 Git으로 굳이 관리할 필요가 없는 특정 파일들이 있어요.
자동으로 생성되는 로그 파일, 빌드 할 때 생기는 컴파일된 파일 등은
`git status` 할 때마다 항상 Untracked 파일이거나 혹은 Modified 상태가 되어
진짜 필요한 정보들을 보기 힘들어질 수도 있어요.
그 외에도 몇 가지 피드백을 더 드렸으니 한번 생각해보세요.
빠른 진행을 위해 Merge 하겠습니다!
